### DISCUSSION TITLE
Is there a place where I can donate to this project?

### DISCUSSION THREAD

**Discussion by noahmartinwilliams** [AccelerateHS/accelerate#528] [General]
I really like this library, but the last update was months ago, and none of the backends currently work for me. Is there a patreon or something that I can make donations to that will speed up the process?

**Comment by tmcdonell** [comment]
At the moment, no. I've thought about this in the past but honestly wasn't sure if anybody would be interested...

I thought that taking an academic job would give me time to continue working on the project, but that hasn't really panned out. If there were enough funding to actually pay somebody (me or otherwise) to dedicate proper (engineering) time to maintaining and developing this project that would be awesome, but I don't know exactly how to get from here to that point \:

Which platforms/backends aren't working for you?

**Reply by noahmartinwilliams** [reply]
Neither the tensorflow nor the llvm backends work on the latest version of Arch Linux. I'm using llvm 15 (which it always complains about), and the tensorflow backend won't even compile.

**Reply by wujilingfeng** [reply]
But this library depends on llvm-9

**Reply by noahmartinwilliams** [reply]
I can't get to llvm 9 without taking a trip through dependency hell.

**Reply by tomsmeding** [reply]
Side note, llvm 12 works using some patched dependencies: https://git.tomsmeding.com/mandelhs/tree/cabal.project (still no llvm 15, though)

**Reply by wujilingfeng** [reply]
`cabal install accelerate --extra-include-dirs= --extra-lib-dirs` can give you some help.

**Reply by tomsmeding** [reply]
Accelerate is a library, and `cabal install` on a library very rarely does something useful, and regularly does something harmful. (At least, for cabal >=3.) In this case, use a cabal.project file with an `extra-lib-dirs:` directive [documentation](https://cabal.readthedocs.io/en/3.8/cabal-project.html#cfg-field-extra-lib-dirs).

**Comment by noahmartinwilliams** [comment]
Ok, an update: I managed to get llvm-12 working and the library. It turns out you need to install llvm-12 in /usr/local (for Linux) and you need to download the llvm-hs repo from github and change the shared-llvm flag to False in the llvm-hs/llvm-hs.cabal file, and then you need to add the Paths modules from each accelerate-llvm sub-project to the autogen-modules section. Hopefully you can see why it took so long to figure some of that out considering I'm still kind of new to Haskell.

**Reply by tmcdonell** [reply]
Indeed, that process is too difficult \\:

**Comment by noahmartinwilliams** [comment]
How much money would you need to work on this full time? I'd like to help out in any way I can. I've got plenty of free time right now, so if I can familiarize myself with the code a bit more I can maybe help make some improvements on it.

**Reply by wujilingfeng** [reply]
Haskell needs more attention and support. Haskell language needs more excellent libraries and developers, and educational institutions should also offer Haskell courses.

**Reply by tmcdonell** [reply]
To work on this full time? My academic salary isn't that great (nothing compared to industry tech salaries) but still probably a lot more than we could get with donations. But I'd also consider other options outside pure crowd funding; working at Haskell Foundation/in industry on this (and other Haskell projects/tooling/etc.) for example.

**Reply by noahmartinwilliams** [reply]
I'm planning on working on a project in the coming months and I might be able to get a source of funding for this. It would be nice to  have some idea of how much it would cost to get accelerate, accelerate-llvm, and it's various dependencies to consistently work.

**Reply by tomsmeding** [reply]
You're right that a new, working release would be a good idea. In fact, we're hoping to release an update to accelerate{,-llvm} before the summer that includes bugfixes from master, plus a refactor to use `clang` instead of linking to LLVM. (The latter allows 1. installing only `clang` instead of the particular LLVM build that the old version needed, and 2. using the newest LLVM version becomes trivial.) At a later date we plan to release a bigger update that reimplements much of the backends and brings various efficiency improvements.

Having said that, while accepting funding for this work may be complicated administratively, maybe we should just have a video call to discuss what your and our plans are and how we may be able to collaborate. Would you be open to a call? If so, do let us know.

**Reply by noahmartinwilliams** [reply]
Hey, sorry I forgot about this. I'd like to do a zoom call with my friend as well since we're planning on having him handle the business side of things.

**Reply by tomsmeding** [reply]
Let's discuss further over email if we're going to discuss specifics; you can reach me [here](mailto:t.j.smeding@uu.nl). For us, the most practical dates in the near future for a video call would be between July 11th and 25th, but later dates are also possible.

**Reply by noahmartinwilliams** [reply]
I sent you an email but I never heard back from you. Did you see it?

**Reply by tomsmeding** [reply]
I... have never seen anything come in, not sure what went wrong! My apologies. Perhaps try again with [this one](mailto:x528@tomsmeding.com)?
